pennylane/qnode.py

```diff
@@ -61,6 +61,51 @@ def pop_jacobian_kwargs(kwargs):
     return circuit_kwargs
 
 
+def _decompose_queue(ops, device):
+    """Recursively loop through a queue and decompose
+    operations that are not supported by a device.
+
+    Args:
+        ops (List[~.Operation]): operation queue
+        device (~.Device): a PennyLane device
+    """
+    new_ops = []
+
+    for op in ops:
+        if device.supports_operation(op.name):
+            new_ops.append(op)
+        else:
+            with OperationRecorder() as rec:
+                op.decomposition(*op.params, wires=op.wires)
+
+            decomposition = _decompose_queue(rec.operations, device)
```

Comment:
Wouldn't it be simpler if `op.decomposition` just returns a list of `Operator` instances?

Reply:
I've changed this to the following:
```python
    @staticmethod
    def decomposition(theta, phi, lam, wires=None):
        ops = [
            Rot(lam, theta, -lam, wires=wires),
            PhaseShift(lam, wires=wires),
            PhaseShift(phi, wires=wires)
        ]
        return ops
```

Reply:
(note that, due to the queuing system, the above works both as a template, _and_ removes the need for the operation recorder, so it is quite nice